browser/ui/ai_chat/print_preview_extractor_unittest.cc

```diff
@@ -29,26 +255,485 @@ class PrintPreviewExtractorTest : public ChromeRenderViewHostTestHarness {
+  }
+
+  template <typename T>
+  void RunTestCase(base::test::TestFuture<T>& future,
+                   const std::string& mime_type,
+                   bool use_capture_pdf,
+                   const std::string& expected_error_msg,
+                   bool extractor_error,
+                   bool simulate_partial_composition) {
+    const bool expect_error = !expected_error_msg.empty();
+    if (!expect_error || extractor_error) {
+      SimulateFullFlow(extractor_error);
+    } else if (simulate_partial_composition) {
+      // Simulate partial composition without OnCompositeToPdfDone
+      internal_extractor()->OnCompositePdfPageDone(
+          0, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+          CreatePageRegion(8).region);
+      internal_extractor()->OnCompositePdfPageDone(
+          1, 0, 0, printing::mojom::PrintCompositor::Status::kSuccess,
+          CreatePageRegion(16).region);
+      internal_extractor()->OnPreviewReady();
+    } else {
+      // For missing preview data case, just trigger preview ready
+      internal_extractor()->OnPreviewReady();
+    }
+
+    auto result = future.Take();
+    if (expect_error) {
+      ASSERT_FALSE(result.has_value())
+          << "Expected error for mime_type=" << mime_type
+          << ", use_capture_pdf=" << use_capture_pdf
+          << ", partial_composition=" << simulate_partial_composition;
+      EXPECT_EQ(result.error(), expected_error_msg);
+    } else {
+      ASSERT_TRUE(result.has_value())
+          << "Expected success for mime_type=" << mime_type
+          << ", use_capture_pdf=" << use_capture_pdf;
+      if constexpr (std::is_same_v<T, ImageResult>) {
+        const std::vector<std::vector<uint8_t>> expected = {{0xde, 0xea},
+                                                            {0xbe, 0xef}};
+        EXPECT_EQ(result.value(), expected);
+      } else {
+        EXPECT_EQ(result.value(), std::string("extracted text"));
+      }
+    }
+  }
+
+  PrintPreviewExtractorInternal* internal_extractor() {
+    return static_cast<PrintPreviewExtractorInternal*>(
+        pp_extractor_->extractor_.get());
+  }
+
  protected:
-  std::unique_ptr<PrintPreviewExtractor> extractor_;
+  std::unique_ptr<PrintPreviewExtractor> pp_extractor_;
 };
 
 TEST_F(PrintPreviewExtractorTest, CapturePdfWithNotPdf) {
-  ASSERT_NE(web_contents()->GetContentsMimeType(), "application/pdf");
+  content::WebContentsTester::For(web_contents())
+      ->SetMainFrameMimeType("text/html");
   base::test::TestFuture<
       base::expected<std::vector<std::vector<uint8_t>>, std::string>>
       future;
-  extractor_->CapturePdf(future.GetCallback());
+  pp_extractor_->CapturePdf(future.GetCallback());
   auto result = future.Take();
   ASSERT_FALSE(result.has_value());
   EXPECT_EQ(result.error(), "Not pdf content");
 }
 
+TEST_F(PrintPreviewExtractorTest, PrintSettings) {
+  // Test with non-PDF content using Extract()
+  RunPrintSettingsTest("text/html", false, true);
```

Comment:
nit: Not specific to this line, I found using Run*Test with most checks (expect) inside the helper function harder to read what the test is about and what do we expect in the test, it would probably be nicer to put EXPECT statements here instead, up to you though.

Reply:
It's for deduplicating the test code, each run case has scoped trace so if some cases failed we can easily tell which line. Added comments for what each run case does in https://github.com/brave/brave-core/pull/29169/commits/e7ec273b7c8e0dfbbd90921c84e530e41f3f47ff